### PR TITLE
REGR: Categorical with np.str_ categories

### LINKED ISSUE #31499: Pandas 1.0 no longer handles `numpy.str_`s as catgories
#### Code Sample

```python
import pandas as pd
pd.Categorical(['1', '0', '1'], [np.str_('0'), np.str_('1')])
```

```pytb
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/angerer/Dev/Python/venvs/env-pandas-1/lib/python3.8/site-packages/pandas/core/arrays/categorical.py", line 385, in __init__
    codes = _get_codes_for_values(values, dtype.categories)
  File "/home/angerer/Dev/Python/venvs/env-pandas-1/lib/python3.8/site-packages/pandas/core/arrays/categorical.py", line 2576, in _get_codes_for_values
    t.map_locations(cats)
  File "pandas/_libs/hashtable_class_helper.pxi", line 1403, in pandas._libs.hashtable.StringHashTable.map_locations
TypeError: Expected unicode, got numpy.str_
```

#### Problem description
I know that having a list of `numpy.str_`s seems weird, but it easily happens when you use non-numpy algorithms on numpy arrays (e.g. `natsort.natsorted` in our case), or via comprehensions or so:

```py
>>> np.array(['1', '0'])[0].__class__
<class 'numpy.str_'>
>>> [type(s) for s in np.array(['1', '0'])]
[<class 'numpy.str_'>, <class 'numpy.str_'>]
```
#### Expected Output
A normal pd.Categorical

#### Pandas version

pandas 1.0

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #31499
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.0.1.rst [hunk 1]
```diff
@@ -19,7 +19,7 @@ Bug fixes
 
 Categorical
 ^^^^^^^^^^^
-
+- Bug in :class:`Categorical` construction with ``numpy.str_`` categories (:issue:`31499`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.1.rst]:
```suggestion
- Fixed regression in :class:`Categorical` construction with ``numpy.str_`` categories (:issue:`31499`)
```

**Review by TomAugspurger — APPROVED** [review]
LGTM. Is this something we want to support long-term?

**Review by jreback — APPROVED** [review]

File: pandas/tests/arrays/categorical/test_constructors.py [hunk 2]
```diff
@@ -408,6 +408,11 @@ def test_constructor_str_unknown(self):
         with pytest.raises(ValueError, match="Unknown dtype"):
             Categorical([1, 2], dtype="foo")
 
+    def test_constructor_np_strs(self):
+        # GH#31499 Hastable.map_locations needs to work on np.str_ objects
+        cat = pd.Categorical(["1", "0", "1"], [np.str_("0"), np.str_("1")])
```
Comment (jreback) [inline pandas/tests/arrays/categorical/test_constructors.py]: actually, I think we should be santizing these inputs on Index construction; we already *do* this for Series IIRC (deep in block manager i think).  Can you see if that's better (as its a more maintainable long term soln). Not averse to your change, but that's *after* we already saved the inputs.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
i would prefer to actually fix this rather than work around np.str_ which is not used anywhere else.

**Comment by TomAugspurger** [comment]
What's the proposed fix?

IMO, we should deprecate the old behavior first if it isn't too costly. This PR doesn't seem too bad, though I don't really understand all its implications.

**Comment by jorisvandenbossche** [comment]
>  I think we should be santizing these inputs on Index construction; we already do this for Series IIRC 

I thought as well, but that doesn't seem to be the case:

```
In [1]: s = pd.Series([np.str_("0"), np.str_("1")])  

In [2]: s[0]        
Out[2]: '0'

In [3]: type(s[0])  
Out[3]: numpy.str_
```

I would keep a possible sanitation for a separate issue / discussion. This PRs seems an easy fix to address the regression.

**Comment by jreback** [comment]
this is just an unsupportable bandaid. this is handling a leak of np.str_ into the internals, which is really bad. I don't think its worth trying to fix this for 1.0.1 like this, rather address a systematic real fix.

**Comment by jorisvandenbossche** [comment]
> this is just an unsupportable bandaid. this is handling a leak of np.str_ into the internals, which is really bad. 

It is what we did before for years. So I don't think it is that unsupportable. 

Since we can still store numpy strings in a Series, and since we supported converting those to a Categorical before, I think this is a good fix. 

For 1.1, we can discuss further if we want to keep supporting this, or want to deprecate it, or want to sanitize on input (to avoid needing to support it).

**Comment by TomAugspurger** [comment]
> For 1.1, we can discuss further if we want to keep supporting this, or want to deprecate it, or want to sanitize on input (to avoid needing to support it).

Agreed.

**Comment by jreback** [comment]
these last minute patches are just causing more and more issues. do what you will this.

**Comment by TomAugspurger** [comment]
@jbrockmendel are you aware of any maintenance burdens or ambiguities this might cause?

At least on 0.25.3, we seem to treat `str` and `np.str_` equivalently in methods like `get_loc` and `unique`

**Comment by jbrockmendel** [comment]
> are you aware of any maintenance burdens or ambiguities this might cause?

Not especially.  If we want to root np.str_ out entirely, that would be a pretty big endeavor.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by TomAugspurger** [comment]
OK. Naively, I agree that rooting out `np.str_` objects in our constructors sounds difficult. At the least, it would require a scan over the values and a couple isinstance checks on each value, which I'd like to avoid if it's not causing problems elsewhere. I'm sure we'll get reports if it is.

**Comment by TomAugspurger** [comment]
Thanks @jbrockmendel.

**Pull request merged**